Add `label clone` command

heaths opened the pull request:
Resolves #5441

heaths commented:
@samcoe if you take #5451 I'll probably refactor into a single package to reuse many of the same private funcs much like the `codespaces` command does. Could use list and then create in a loop which will automatically take advantage of the `--force` switch support I added.

samcoe commented:
@heaths I think #5451 is a good addition and it makes sense to extract out common functionality into a `shared` or `common` file which we do in a couple different commands 👍

heaths commented:
I could do that as well. The funcs would have to be public, of course, but all that is under internal/ anyway.

heaths committed:
Flatten the label package

Keeps shared functions privage to the package while sharing with all label commands.

heaths committed:
Add `label clone` command

heaths commented:
@samcoe wanted to get your preliminary feedback on this. Once I incorporate any feedback I will add tests and convert this to a real PR.

samcoe commented on api/queries_repo.go:
```diff
@@ -215,6 +215,15 @@ func (r Repository) RepoHost() string {
 }
 
+// String returns the [HOST/]OWNER/NAME of the repository
+func (r Repository) String() string {
```
I don't know why we have waited so long to add this, I have wanted to do this for awhile. I don't think we should output the hostname though. As you can see from where the `ghrepo.FullName` helper, in most of the cases that require serializing a `ghrepo` we just want the `OWNER/NAME` format. After this change I would be inclined to change `ghrepo.FullName` to now include the hostname for when the fully qualified name is necessary.

heaths replied:
To clarify, you're saying that `String()` should never emit a hostname, but `FullName()` should if defined?

One thing I did notice is that, in practice, hostname never seems to be empty. When testing with some scratch repos I created, I'd get something like "github.com/heaths/test1".

samcoe replied:
For now lets have `String()` be a clone of `FullName()` and then in the future we can transition `FullName()` to include the hostname. That transition feels like scope creep and can be done at a later time though. 

Hostname is never going to be empty, unless explicitly set to `""`. We set it to `ghinstance.Default()` in all other cases I believe.

samcoe commented on internal/ghrepo/repo.go:
```diff
@@ -144,3 +145,11 @@ func (r ghRepo) RepoName() string {
 	return r.hostname
 }
+
+func (r ghRepo) String() string {
```
Same comment as the above `String()` function.

samcoe commented on pkg/cmd/label/clone.go:
```diff
@@ -0,0 +1,141 @@
+			current repo using the --force flag.
+
+		`),
+		Example: heredoc.Doc(`
```
An example specifying how to clone between two remote repos I think would be helpful.
```
gh label clone cli/cli -R samcoe/cli
```

heaths replied:
Yep. 🙂 I'll make sure docs are complete/improved and tests added before promoting this PR from a draft.
gh label clone

samcoe commented on pkg/cmd/label/clone.go:
```diff
@@ -0,0 +1,141 @@
+		Use:   "clone <source [HOST/]OWNER/REPO>",
+		Short: "Clones labels from another repo",
+		Long: heredoc.Doc(`
+			Clones labels from another repo into the current repo on GitHub.
```
I think referring to this as cloning labels from a source repo into a destination repo, and specifying that the current repo is the default destination repo and can be overridden using the standard `-R` flag is more clear.

samcoe replied:
This allows us to use "destination repo" instead of "current repo" in all documentation which feels more accurate.

samcoe commented on pkg/cmd/label/clone.go:
```diff
@@ -0,0 +1,141 @@
+	}
+
+	cmd := &cobra.Command{
+		Use:   "clone <source [HOST/]OWNER/REPO>",
```
For consistency with other commands lets use `repository` rather than the accepted format. 
```suggestion
		Use:   "clone <source-repository>",
```

heaths replied:
Something like that was my original thought, but my worry was that there's no way to specify documentation for arguments; though, I suppose I could just put something in the description. I wanted to make it clear that they could use the same "[host/]owner/repo" that they can for `-R`.

Or do you think samples alone would be enough?

Long-term since you're already overriding doc output for cobra anyway, should arguments get their own documentation? `gh` uses quite a few arguments and some could benefit users- especially new users - with more description.

samcoe replied:
I think that most our users are familiar with the format we accept for repo arguments, and if they do enter an incorrect format the error message displays the formats we accept. Additionally an example or two should be sufficient enough that we don't need to add anything to the description explicitly.

I think arguments getting their own documentation sounds like good idea and perhaps something we can address in the future.

samcoe commented on pkg/cmd/label/clone.go:
```diff
@@ -0,0 +1,141 @@
+		},
+	}
+
+	cmd.Flags().BoolVarP(&opts.Force, "force", "f", false, "Overwrite labels in the current repo from the source repo.")
```
"Overwrite labels in the destination repository" seems like it is descriptive enough.

samcoe commented on pkg/cmd/label/clone.go:
```diff
@@ -0,0 +1,141 @@
+		return err
+	}
+
+	successCount, totalCount, err := cloneLabels(opts)
```
In non-tty mode, for scripts, what do you think the behavior should be when not all of the labels are successfully cloned? Should that result in an error exit code or a successful one?

heaths replied:
Good question. Since `--force` is defined to specifically overwrite and cloning really is more of a "merging" without it, I'm inclined to *not* exit with an error code unless we failed completely...

...though, now that I opine on that, perhaps we shouldn't fail ever (even with the TTY messages) if the only error was that a duplicate name was already defined.

Thoughts?

samcoe replied:
Yeah I agree, I think that behavior makes sense. Only exit with error status if we receive an error that is not "label already exists".

samcoe commented on pkg/cmd/label/clone.go:
```diff
@@ -0,0 +1,141 @@
+	defer opts.IO.StopProgressIndicator()
+
+	successCount = 0
+	labels, totalCount, err := listLabels(httpClient, opts.SourceRepo, listLimit)
```
What happens if there are more than the `listLimit` number of labels? Seems like we should check first how many labels there are and then use that here rather than hardcode the `listLimit`.

heaths replied:
I admit I didn't scan through enough of `listLabels`. Initially it seemed like the limit was just a page size. I see now it's actually a hard limit. IMO, this shouldn't be in `label list`. Why not just list all labels using pageable results like I do in heaths/gh-label? I don't see a customer value here, especially labels are not returned in any documented deterministic order.

IMO, we should actually eliminate the default value of 30 and not use a limit if not specified explicitly. I.e. page from `listLabels`. The limit could be made optional - if not in the GraphQL query (if supported), we can use separate GraphQL queries for limited vs. paged.

heaths replied:
To clarify, it is used as the page size and does support paging, but lines 67-69 of this block of code should be removed:

https://github.com/cli/cli/blob/c99c47f34db5d1ff742ae2bb89a64d7b6fb2aa06/pkg/cmd/label/list/http.go#L65-L70

Happy to do that in this PR or another, but I don't see any customer value in this. For issues, a limit makes sense because there are additional query parameters you can specify. Nothing like that exists for labels, but maybe it should. My [extension](https://github.com/heaths/gh-label#list) does for use with a repo that has a lot of labels e.g. [Azure/azure-sdk-for-net](https://github.com/Azure/azure-sdk-for-net). Even then, seems a limit should not be used by default, especially given the non-deterministically unordered results.

Adding an optional filter for labels would not be a breaking change anyway - not practically, anyway. IF someone passed an invalid arg expecting an error, this may now work but that doesn't seem like a valid test (outside of functional tests for this repo) anyway. Could always add another parameter like `--search` just like `issue list` has, too.

Speaking of unsorted results, may as well sort as well. Not as useful if hard to visually scan.

heaths replied:
Opened #5489 to track.

samcoe replied:
IMO the customer value provided by having the limit is two fold. 

First the UX is better with shorter lists on the terminal, 30 items generally fit on a single terminal screen and doesn't take up too much terminal height. It is also consistent with our other list commands which I believe use 30 as a default. The UX argument is subjective, and can be combated by starting up the pager process when outputting the label list. 

The second, is that it keeps the user from potentially burning through their API rate limit inadvertently. Likely, the couple extra requests needed in most cased to retrieve all the repo's labels won't greatly affect their rate limit but I think it is nice to add a bit of friction to get the user to think a bit about what they are doing by setting a large limit rather than implicitly running any number of queries to get all the labels. 

Additionally, the labels are actually sorted by created at date by default when returned from the API so they are determanistic. I am open to expanding the feature set of the `label list` command to include ordering and searching as the API supports those but feels like scope creep to add in here and should be addressed in a separate PR.

samcoe replied:
In order to get `clone` working and put off major `list` work till another time I think it would make sense to pass in `-1` or perhaps `0` into `listLabels` as the limit and that function know that means to keep retrieving until there are no more pages.

heaths replied:
I'll follow up after this PR with another for `label list` (and `label remove`) but I'll lay the groundwork for paging as you suggest.

samcoe commented on pkg/cmd/label/http.go:
```diff
@@ -1,14 +1,13 @@
 )
 
-func listLabels(client *http.Client, repo ghrepo.Interface, limit int) ([]shared.Label, int, error) {
+func listLabels(client *http.Client, repo ghrepo.Interface, limit int) ([]label, int, error) {
```
Is there a good reason not to just move `listLabels` and `createLabels` to the already existing `shared` package? I think that path conforms with what we have done with other commands but am open to hearing reasons why this is a might be a better solution.

heaths replied:
The codespaces package/command uses the pattern I did here. IMO, making more public than one has to opens up more maintenance hassles. Granted, this module is intended to create an executable, but I've more typically seen projects start from the least permissive accessibility required and open up as needed.

But that was just my reasoning. I can change it back and make what I need public out of the `shared` package if you prefer.

samcoe replied:
I don't have a strong preference and I think it is reasonable to keep as much non-exported as possible.

samcoe reviewed (commented):
@heaths This is coming along nicely. I left a handful of comments but nothing major I don't think. Let me know if you have any questions.

heaths committed:
Resolve PR feedback

heaths committed:
Fetch all labels

heaths committed:
Use consistent help wording

samcoe committed:
Small amount of polish and cleanup

samcoe reviewed (approved):
@heaths Appreciate the hard work here. This looks great, thanks for the thorough tests. I pushed a small commit to polish up a couple tiny things rather than going through another round of review. I will get this merged before the next release.

samcoe commented:
Some thoughts for future improvements that we might want to address after testing out the command:
1. Parallelizing the creation of labels. Even on a small amount of labels (40) the command can take seconds to run. 
2. Adding in a confirmation prompt, generally we reserve this for destructive actions, but I can see how it would be valuable since it can be super tedious to undo (deleting) numerous labels.

I don't think either needs to be addressed right away just food for thought.

heaths commented:
> I will get this merged before the next release.

@samcoe, anything holding it up now? I have another PR that is based on this work to improve the `label list` command and add the `label remove` command, but didn't want to make it a full PR until I can rebase onto trunk from where it diverged this with one. IMO, making the label command complete by next release would be ideal.

> Parallelizing the creation of labels. Even on a small amount of labels (40) the command can take seconds to run.

I've never actually tested this nor seen it in docs, but is there a simultaneous connection limit per client ID for GitHub APIs? And, if so, is it different for REST vs. GraphQL? This could significantly impact the design. Could either go with a fixed-sized goroutine pool or just span goroutines as needed e.g. per label creation/update. Normally I'd lean toward the former.

heaths commented on pkg/cmd/label/clone_test.go:
```diff
@@ -392,59 +392,87 @@ func TestCloneRun(t *testing.T) {
-							Color: "00ff00",
+				reg.Register(
+					httpmock.GraphQL(`query LabelList\b`),
+					httpmock.StringResponse(`
```
@samcoe any reason we wouldn't want to also verify the variable inputs? This decreases coverage by ignoring request assertions.

samcoe replied:
No, just missed it when making changes. Thanks for pointing it out. It should be fixed now.

heaths commented on pkg/cmd/label/http.go:
```diff
@@ -79,9 +80,10 @@ loop:
-		return a
+func determinePageLimit(limit int) int {
+	// If limit is -1 we want to retrieve all labels
+	if limit < 0 || limit > defaultPageLimit {
```
@samcoe this seems antithetical to your previous statement about throttling. Users will never be able to request more than 30 items which means more requests for repos with more than 30 labels. Why not allow a page size up to 100 which, IIRC, is enforced as the max in the service anyway? It's the chatty-vs-chunky decision.

samcoe replied:
I am not sure I am following. I am confident the current logic is correct but the variable names I have used are making it confusing as to what is happening here. `defaultPageLimit` should be renamed `maxPageSize` and `limit` should be renamed `numRequestedItems` or similar. This function should return the `maxPageSize` if the user requests more than the `maxPageSize` or if the user has requested `-1`, otherwise return the `numRequestedItems`.

heaths replied:
Understood, though I must admit I missed that the `defaultPageLimit` was changed to 100 from 30. 😳

samcoe committed:
Address feedback

heaths commented:
Rebased on trunk to resolve merge conflicts.

heaths commented:
@samcoe here the other one now that I've rebased it onto trunk: https://github.com/cli/cli/pull/5503

adriens commented:
Documented in https://cli.github.com/manual/gh_label_clone
Very useful :heavy_heart_exclamation:  Thanks a lot.

It helps on https://github.com/github/feedback/discussions/15300

adriens commented:
:gift: Thanks a lot for the feature :pray: 
Here is my feedback : https://dev.to/adriens/github-cli-v290-is-out-clone-labels-feature-5452-is-here-3odk
